propagators/opentelemetry-propagator-aws-xray/src/AWSXRayPropagator.ts

```diff
@@ -0,0 +1,200 @@
+
+const TRACE_ID_KEY = 'Root';
+const TRACE_ID_LENGTH = 35;
+const TRACE_ID_VERSION = "1";
+const TRACE_ID_DELIMITER = '-';
+const TRACE_ID_DELIMITER_INDEX_1 = 1;
+const TRACE_ID_DELIMITER_INDEX_2 = 10;
+const TRACE_ID_FIRST_PART_LENGTH = 8;
+
+const PARENT_ID_KEY = 'Parent';
+// const PARENT_ID_LENGTH = 16;
+
+const SAMPLED_FLAG_KEY = 'Sampled';
+const SAMPLED_FLAG_LENGTH = 1;
+const IS_SAMPLED = '1';
+const NOT_SAMPLED = '0';
+
+const VALID_TRACEID_REGEX = /^([0-9a-f]{32})$/i;
+const VALID_SPANID_REGEX = /^[0-9a-f]{16}$/i;
+/**
+ * Implementation of the AWS X-Ray Trace Header propagation protocol. See <a href=
+ * https://https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader>AWS
+ * Tracing header spec</a>
+ * 
+ * An example AWS Xray Tracing Header is shown below:
+ * X-Amzn-Trace-Id: Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1
+ */
+export class AWSXRayPropagator implements HttpTextPropagator {
+  inject(context: Context, carrier: unknown, setter: SetterFunction) {
+    const spanContext = getParentSpanContext(context);
+    if (!spanContext) return;
+    if (!VALID_TRACEID_REGEX.test(spanContext.traceId) ||
+      !VALID_SPANID_REGEX.test(spanContext.spanId) ||
+      (spanContext.traceFlags !== TraceFlags.NONE &&
+        spanContext.traceFlags !== TraceFlags.SAMPLED)) return;
+
+    const otTraceId = spanContext.traceId;
+    const xrayTraceId = TRACE_ID_VERSION
+      + TRACE_ID_DELIMITER
+      + otTraceId.substring(0, TRACE_ID_FIRST_PART_LENGTH)
+      + TRACE_ID_DELIMITER
+      + otTraceId.substring(TRACE_ID_FIRST_PART_LENGTH);
+    const parentId = spanContext.spanId;
+    const samplingFlag = spanContext.traceFlags? IS_SAMPLED: NOT_SAMPLED;
+    // TODO: Add OT trace state to the X-Ray trace header
+
+    const traceHeader = TRACE_ID_KEY
+      + KV_DELIMITER
+      + xrayTraceId
+      + TRACE_HEADER_DELIMITER
+      + PARENT_ID_KEY
+      + KV_DELIMITER
+      + parentId
+      + TRACE_HEADER_DELIMITER
+      + SAMPLED_FLAG_KEY
+      + KV_DELIMITER
+      + samplingFlag;
+
+    setter(
+      carrier,
+      AWSXRAY_TRACE_ID_HEADER,
+      traceHeader
+    );
+  }
+
+  extract(context: Context, carrier: unknown, getter: GetterFunction): Context {
+    const spanContext = this.getSpanContextFromHeader(carrier, getter);
+    if (!isValid(spanContext)) return context;
+
+    return setExtractedSpanContext(context, spanContext);
+  }
+  
+  private getSpanContextFromHeader(carrier: unknown, getter: GetterFunction): SpanContext {
+    const traceHeader = getter(carrier, AWSXRAY_TRACE_ID_HEADER);
+    // Only if the returned traceHeader is no empty string can be extracted
+    if (!traceHeader || typeof traceHeader !== 'string') return INVALID_SPAN_CONTEXT;
+
+    let pos = 0, trimmedPart: string;
+    let parsedTraceId = INVALID_TRACE_ID;
+    let parsedSpanId = INVALID_SPAN_ID;
+    let parsedTraceFlags = null;
+    while(pos < traceHeader.length) {
```

Comment:
In Java, this is quite a bit faster than using `split`, but I don't know about JS, it may be better to stick to using `split` here.

Reply:
I am also confused here, I initially tried to use split method but the Java implementation is in this way so I just followed it

Reply:
There is a benchmark utility you can use to test things like this. You can see an example at https://github.com/open-telemetry/opentelemetry-js/issues/1334

You can also use jsperf.com for browser benchmarking.

Reply:
Subjectively, I think `split` based implementations are typically easier to read, which is an important distinction. I would expect your implementation here to be slightly faster than a split implementation, but not appreciably.

Reply:
If the ids are fixed length, it might be easier and faster to use substring also.

Reply:
I now think Daniel's consideration makes more sense. Since we have a fixed size tracing header, it can be a little bit faster to use substring, and maybe that is why Java also implements in this way